### DISCUSSION TITLE
How to obtain the transformation sparse arrays for 3D cone beam or parallel beam projection?

### DISCUSSION THREAD

**Discussion by EYUJin** [astra-toolbox/astra-toolbox#431] [Q&A]
Hi,

I'm trying to obtain transformation sparse arrays for 3D cone beam/parallel beam projections, similar to the issues:
#115, #203 

I understand that the projection matrices for 3D CT are computed on-the-fly when needed.

However, I don't get how I could implement the following (the work-around mentioned in #203 ):

"You can of course form it explicitly one column at a time by projecting unit vectors."

Could you provide examples for the solution above for cone beam and parallel beam configurations respectively?

The reason this is quite perplexing is because according to...

```
def weights_single_ray(i, projection_index, detector_index):
    raise NotImplementedError("Not yet implemented")

def weights_projection(i, projection_index):
    raise NotImplementedError("Not yet implemented")
```

(https://github.com/astra-toolbox/astra-toolbox/blob/7273498d63451ee75c5389b693fe301f9f04d356/python/astra/projector3d_c.pyx#L109C1-L114C53)

the functions weights_single_ray and weights_projection are not implemented for the 3d projector items.

I'd really appreciate your reply.

**Comment by askorikov** [comment]
Here is an example code:

```python
import astra
import numpy as np
from tqdm import trange
import scipy.sparse
import matplotlib.pyplot as plt


N = 32
N_ANGLES = 32

angles = np.linspace(0, 2*np.pi, N_ANGLES)
proj_geom = astra.create_proj_geom('cone', 1.0, 1.0, N, N, angles, 2*N, N)
vol_geom = astra.create_vol_geom(N, N, N)

proj_matrix_cols = []
for i in trange(N**3):
    unit_vector = np.zeros(N**3, dtype=np.float32)
    unit_vector[i] = 1.0
    unit_vector = unit_vector.reshape(N, N, N)
    _, proj = astra.create_sino3d_gpu(unit_vector, proj_geom, vol_geom)
    proj_matrix_col = proj.reshape(N*N_ANGLES*N, 1)
    proj_matrix_col = scipy.sparse.csr_array(proj_matrix_col)
    proj_matrix_cols.append(proj_matrix_col)
proj_matrix = scipy.sparse.hstack(proj_matrix_cols)

# Test on a cube-shaped phantom
phantom = np.zeros([N, N, N])
phantom[N//4:-N//4, N//4:-N//4, N//4:-N//4] = 1.0
proj = (proj_matrix @ phantom.flatten()).reshape(N, N_ANGLES, N)
plt.imshow(proj[:, N_ANGLES//3, :])
```

Note that it takes a lot of time to calculate and a lot of memory to store the resulting matrix for realistic sizes of data.

**Comment by ChemEng227** [comment]
You could try wheels built by Pylops package. With simply
```
W = astra.OpTomo(projector_id)
W_lops = pylops.LinearOperator(W)
W_np = W_lops.tosparse()
```
You can get the weights in scipy sparse matrix form.
